src/freenet/node/AnnounceSender.java

```diff
@@ -484,6 +484,15 @@ private void timedOut(PeerNode next) {
 	}
 
 	private void rnf(PeerNode next) {
+		synchronized(this) {
+			while(waitingForTransfers > 0) {
+				try {
+					wait();
+				} catch (InterruptedException e) {
+					// Ignore.
```

Comment:
Deduplication would be nice, indeed. The hard part is coming up with a decent way to handle that exception. :)